apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex

```diff
@@ -0,0 +1,377 @@
+       call_stack: call_stack
+     }
+   end
+
+   def init(%{
+         contract: contract,
+         vm_version: vm_version,
+         amount: amount,
+         gas: gas,
+         gas_price: gas_price,
+         call_data: call_data,
+         call_stack: call_stack
+       }) do
+     identified_contract = Identifier.create_identity(contract, :contract)
+
+     %ContractCallTx{
+       contract: identified_contract,
+       vm_version: vm_version,
+       amount: amount,
+       gas: gas,
+       gas_price: gas_price,
+       call_data: call_data,
+       call_stack: call_stack
+     }
+   end
+
+   @spec validate(t(), DataTx.t()) :: :ok | {:error, String.t()}
+   def validate(
+         %ContractCallTx{
+           contract: contract
+         },
+         _data_tx
+       ) do
+     if validate_identifier(contract, :contract) do
+       :ok
+     else
+       {:error, "#{__MODULE__}: Invalid contract address: #{inspect(contract)}"}
+     end
+   end
+
+   @spec process_chainstate(
+           Chainstate.accounts(),
+           tx_type_state(),
+           non_neg_integer(),
+           t(),
+           DataTx.t(),
+           Transaction.context()
+         ) :: {:ok, {Chainstate.accounts(), tx_type_state()}}
+   def process_chainstate(
+         accounts,
+         calls,
+         block_height,
+         %ContractCallTx{} = call_tx,
+         data_tx,
+         context
+       ) do
+     # Transfer the attached funds to the callee, before the calling of the contract
+     sender = DataTx.main_sender(data_tx)
+     nonce = DataTx.nonce(data_tx)
+
+     chain_state = Chain.chain_state()
+
+     updated_accounts_state =
+       accounts
+       |> AccountStateTree.update(sender, fn acc ->
+         Account.apply_transfer!(acc, block_height, call_tx.amount)
+       end)
+
+     updated_chain_state = Map.put(chain_state, :accounts, updated_accounts_state)
+
+     init_call = Call.new(sender, nonce, block_height, call_tx.contract.value, call_tx.gas_price)
+
+     {call, update_chain_state1} =
+       run_contract(call_tx, init_call, block_height, nonce, updated_chain_state)
+
+     accounts1 = update_chain_state1.accounts
+
+     accounts2 =
+       case context do
+         :contract ->
+           accounts1
+
+         :transaction ->
+           gas_cost = call.gas_used * call_tx.gas_price
+           amount = DataTx.fee(data_tx) + gas_cost
+           caller1 = AccountStateTree.get(accounts1, sender)
+
+           accounts1
+           |> AccountStateTree.update(caller1.id.value, fn acc ->
+             Account.apply_transfer!(acc, block_height, amount)
+           end)
+       end
+
+     # Insert the call into the state tree. This is mainly to remember what the
+     # return value was so that the caller can access it easily.
+     # Each block starts with an empty calls tree.
+     updated_calls_tree = CallStateTree.insert_call(calls, call)
+
+     {:ok, %{update_chain_state1 | accounts: accounts2, calls: updated_calls_tree}}
```

Comment:
This doesn't follow `@spec`. Following `@spec` is required by the behaviour every tx has to follow. 

Correct is: `{:ok, {accounts2, updated_calls_tree}}`